fling/utils/attack_utils/dlg.py

```diff
@@ -0,0 +1,245 @@
+        """
+        start, end = 0, batch_size
+        batch_idx = 0
+        total_loss = []
+        total_min_loss = []
+
+        model = model.to(device)
+        model = model.eval()
+        tv_criterion = TVLoss()
+
+        # Get the parameters whose gradients are visible to the attacker.
+        if not isinstance(parameter_args, EasyDict):
+            parameter_args = EasyDict(parameter_args)
+        visible_parameters = get_weights(model, parameter_args)
+
+        # Iterate for each batch in the dataset.
+        while end <= len(dataset):
+            if start == end:
+                break
+            batch = dataset[start:end]
+            start = end
+            end = min(len(dataset), end + batch_size)
+            self.logger.logging(f'Star batch: {batch_idx}...')
+            batch_idx += 1
+
+            # Prepare input data and labels.
+            batch_x = torch.stack([batch[i]['input'] for i in range(len(batch))]).to(device)
+            batch_y = torch.tensor([batch[i]['class_id'] for i in range(len(batch))]).to(device).unsqueeze(1)
+            gt_label = torch.zeros(batch_y.size(0), class_number, device=device)
+            gt_label.scatter_(1, batch_y, 1)
+
+            # Prepare dummy data and label. Set the corresponding optimizer.
+            dummy_data = torch.rand_like(batch_x, requires_grad=True)
+            if use_gt_labels:
+                dummy_label = gt_label
+                if optim_backend == 'lbfgs':
+                    optimizer = torch.optim.LBFGS([dummy_data])
+                elif optim_backend == 'adam':
+                    optimizer = torch.optim.Adam([dummy_data])
+                else:
+                    raise ValueError(f'Unrecognized optimizer: {optim_backend}.')
+            else:
+                dummy_label = torch.randn(batch_y.size(0), class_number, device=device, requires_grad=True)
+                if optim_backend == 'lbfgs':
+                    optimizer = torch.optim.LBFGS([dummy_data, dummy_label])
+                elif optim_backend == 'adam':
+                    optimizer = torch.optim.Adam([dummy_data, dummy_label])
+                else:
+                    raise ValueError(f'Unrecognized optimizer: {optim_backend}.')
+
+            # Calculate the ground truth gradient.
+            gt_pred = model(batch_x)
+            gt_onehot_label = torch.softmax(gt_label, dim=-1)
+            gt_loss = torch.mean(torch.sum(-gt_onehot_label * F.log_softmax(gt_pred, dim=-1), 1))
+            gt_dy_dx = torch.autograd.grad(gt_loss, visible_parameters, create_graph=True, retain_graph=True)
+
+            # This is for saving reconstructed images.
+            batch_img_history = {self.base_index + start + i: [] for i in range(dummy_data.shape[0])}
+
+            # Save the min reconstruction loss and the corresponding images.
+            best_losses = torch.zeros(dummy_data.shape[0], dtype=torch.float).to(device)
```

Comment:
The psnr value may be less than 0, so the `best_losses` should not be initialized as 0.